mlir/include/mlir/Dialect/GPU/IR/GPUOps.td

```diff
@@ -1345,12 +1345,40 @@ def GPU_BarrierOp : GPU_Op<"barrier"> {
   let hasCanonicalizer = 1;
 }
 
+def SpatialExtentsAttr
+  : GPU_Attr< "SpatialExtents", "spatial_extents"> {
+  let summary = "Specify the spatial extent of a gpu module such as subgroup size.";
+  let description = [{
+    Specify the spatial extent of a GPUModuleOp. This can be useful for enabling
+    algorithms, optimizations, or features that rely on a particular kernel
+    parameters being true e.g. vendor specific extentions that only work for one
+    subgroup size. `reqdSubgroupSize` specifies that the module is only valid
+    when subgroup size is that value. `reqdWorkgroupSize` specifies that the
+    module is only valid when workgroup size is that value. `maxWorkgroupSize`
+    specifies that the module is only valid when workgroup size is less than or
+    equal to that value.
+
+    All parameters are optional. Having both `reqdWorkgroupSize` and
+    `maxWorkgroupsize` is invalid.
+  }];
+
+  let parameters = (ins
+  OptionalParameter<"std::optional<unsigned>">: $reqdSubgroupSize,
+  OptionalParameter<"std::optional<DenseI32ArrayAttr>">: $reqdWorkgroupSize,
+  OptionalParameter<"std::optional<DenseI32ArrayAttr>">: $maxWorkgroupSize);
```

Comment:
```suggestion
  OptionalParameter<"unsigned">: $reqdSubgroupSize,
  OptionalParameter<"DenseI32ArrayAttr">: $reqdWorkgroupSize,
  OptionalParameter<"DenseI32ArrayAttr">: $maxWorkgroupSize);
```

`optional^2`

Reply:
I believe `OptionalParameter` is a parameter that will default to the default initialization of the storage type. Not a great name...

Reply:
I see. In that case, we may wanna go with:

```suggestion
  OptionalParameter<"IntegerAttr">: $reqdSubgroupSize,
  OptionalParameter<"DenseI32ArrayAttr">: $reqdWorkgroupSize,
  OptionalParameter<"DenseI32ArrayAttr">: $maxWorkgroupSize);
```

`Attr` can still be nullable and I see this is a more common pattern in MLIR.

Reply:
done